src/compiler/checker.ts

```diff
@@ -6653,10 +6653,98 @@ namespace ts {
 
             return type;
 
+            function narrowTypeByValueExpression(type: Type, expr: BinaryExpression, assumeTrue: boolean): Type {
+                assumeTrue = (expr.operatorToken.kind === SyntaxKind.EqualsEqualsEqualsToken) ? assumeTrue : !assumeTrue;
+                let lhs = skipParenthesizedNodes(expr.left);
+                const selectors: string[] = [];
+                while (lhs.kind !== SyntaxKind.Identifier) {
+                    switch (lhs.kind) {
+                        case SyntaxKind.PropertyAccessExpression:
+                            selectors.push((lhs as PropertyAccessExpression).name.text);
+                            // If a name doesn't resolve, bail
+                            if (selectors[selectors.length - 1] === undefined) {

```

Comment:
Instead of unconditionally pushing and grabbing the last one, use a local variable.